### PR TITLE
Add support for paginated content to entity detection.

### PR DESCRIPTION
**Pull request by techanvil**
## Summary

Addresses issue:

- #1911 

## Relevant technical choices

It was necessary to bring in some code from WordPress core in order to meet the AC.

Notably, the new method `Entity_Factory::paginate_entity_url` contains code derived from the WordPress function [`paginate_links`](https://github.com/WordPress/WordPress/blob/7f5d7f1b56087c3eb718da4bd81deb06e077bbbb/wp-includes/general-template.php#L4203), and `Entity_Factory::paginate_post_url` contains code derived from [`_wp_link_page`](https://github.com/WordPress/WordPress/blob/7f5d7f1b56087c3eb718da4bd81deb06e077bbbb/wp-includes/post-template.php#L1031).

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: includes/Core/Util/Entity_Factory.php [hunk 1]
```diff
@@ -560,4 +568,108 @@ private static function convert_to_amp_entity( $new_url, $entity ) {
...
+	 * Logic extracted from `paginate_links` in WordPress core.
+	 * https://github.com/WordPress/WordPress/blob/7f5d7f1b56087c3eb718da4bd81deb06e077bbbb/wp-includes/general-template.php#L4203
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $url The URL to paginate.
+	 * @param int    $pagenum The page number to add to the URL.
+	 * @return string The paginated URL.
+	 */
+	private static function paginate_entity_url( $url, $pagenum ) {
```
Comment (eugene-manuilov) [inline includes/Core/Util/Entity_Factory.php:597]: How about using something like this:

```php
$original_request_uri = $_SERVER['REQUEST_URI'];
$_SERVER['REQUEST_URI'] = str_replace( home_url(), '', $url );

$paginated_url = get_pagenum_link( $pagenum );

$_SERVER['REQUEST_URI'] = $original_request_uri;

return $paginated_url;
```

It is a bit hacky but does what we need without copy/pasting functionality from WordPress core.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @techanvil. Added two comments for you. Please, take a look.

File: includes/Core/Util/Entity_Factory.php [hunk 2]
```diff
@@ -54,7 +54,7 @@ public static function from_context() {
 
 			$post = get_post();
 			if ( $post instanceof WP_Post && self::is_post_public( $post ) ) {
-				return self::create_entity_for_post( $post );
+				return self::create_entity_for_post( $post, self::get_query_pagenum( $wp_the_query, 'page' ) );
```
[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Core/Util/Entity_Factory.php]: Thanks @eugene-manuilov, that's a good spot! I've updated to always use `1` for the page here.

File: includes/Core/Util/Entity_Factory.php (hunk 1 above)
Comment (techanvil) [reply includes/Core/Util/Entity_Factory.php:597]: Thanks @eugene-manuilov, I would much prefer to reuse existing code too. Unfortunately, I did initially try reusing the logic from `get_pagenum_link`, but discovered in testing that it resulting in mismatched URLs in some cases when using the Plain Permalinks setting.

For example, with a paginated category, having navigated to page 2 via the pagination links, the URL is `http://sitekit.10uplabs.com/?paged=2&cat=3`.

image

But then, navigating via the Admin Bar to the Entity Dashboard:

image

The URL for the page doesn't match, the order of the query params is reversed: https://sitekit.10uplabs.com/?cat=3&paged=2

image

Using the logic from `paginate_links` results in the correct URL being shown in SiteKit. I realise there's a possibility that sometimes `get_pagenum_link` may be used for navigation instead but given the prominence of the paginated navigation I would think that's going to be the most common way for a user to get to a paginated page, although I could of course be mistaken.

**Review by eugene-manuilov — APPROVED** [review]
LGTM

**Pull request merged**